### PR TITLE
[RFR] Pre-commit hooks

### PR DESCRIPTION
**Pull request by djhi**
Follow #3304 
We currently refuse some pull requests because of linting/formatting issues far too often.

This PR leverage `husky` and `lint-staged` to run prettier and tslint on every modified files. Besides, it also run the unit tests related to those files, and only those, ensuring a fast feedback loop without having to wait for a Travis build only to realize you forgot to run the tests for your changes. Thanks Jest!

### CODE REVIEW

**Comment by fzaninotto** [comment]
Can you rebase?

File: Makefile [hunk 1]
```diff
@@ -107,17 +107,8 @@ prettier: ## prettify the source code using prettier
...
-	@if [ "$(CI)" != "true" ]; then \
-		echo "Running unit tests..."; \
-		yarn -s test-unit; \
-	fi
-	@if [ "$(CI)" = "true" ]; then \
-		echo "Running unit tests in CI..."; \
-		yarn -s test-unit-ci; \
-	fi
-
-test-unit-watch: ## launch unit tests and watch for changes
```
Comment (fzaninotto) [inline Makefile]: why did you remove this one?

File: package.json [hunk 2]
```diff
@@ -5,13 +5,11 @@
         "build": "lerna run build",
         "build-demo": "cd examples/demo && cross-env REACT_APP_DATA_PROVIDER=rest yarn -s build",
         "test-unit": "cross-env NODE_ENV=test cross-env BABEL_ENV=cjs NODE_ICU_DATA=node_modules/full-icu jest",
-        "test-unit-ci": "cross-env NODE_ENV=test cross-env BABEL_ENV=cjs NODE_ICU_DATA=node_modules/full-icu jest --runInBand",
```
Comment (fzaninotto) [inline package.json]: why did you remove this one?

File: Makefile (hunk 1 above)
Comment (djhi) [reply Makefile]: See README.

Comment (djhi) [reply Makefile]: Nvm I'll revert that. Jest does not watch automatically when not on CI. That's done by react-scripts.

File: Makefile [hunk 3]
```diff
@@ -117,7 +117,8 @@ test-unit: ## launch unit tests
 	fi
 
 test-unit-watch: ## launch unit tests and watch for changes
-	yarn -s test-unit --watch
+	echo "Running unit tests in CI..."; \
```
Comment (fzaninotto) [inline Makefile]: No, not necessarily in CI

Comment (djhi) [reply Makefile]: Damn copy/paste!
